components/cereal-service/pkg/server/server.go

```diff
@@ -0,0 +1,632 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/gofrs/uuid"
+	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/sirupsen/logrus"
+	"github.com/teambition/rrule-go"
+
+	"github.com/chef/automate/api/interservice/cereal"
+	libcereal "github.com/chef/automate/lib/cereal"
+	"github.com/chef/automate/lib/cereal/backend"
+)
+
+var errInvalidMsg = errors.New("invalid msg")
+
+type CerealService struct {
+	workflowScheduler *libcereal.WorkflowScheduler
+	backend           backend.Driver
+}
+
+func NewCerealService(ctx context.Context, b backend.Driver) *CerealService {
+	cs := &CerealService{
+		backend: b,
+	}
+
+	if v, ok := b.(backend.SchedulerDriver); ok {
+		cs.workflowScheduler = libcereal.NewWorkflowScheduler(v)
+		go cs.workflowScheduler.Run(ctx)
+	}
+
+	return cs
+}
+
+func generateRequestID() string {
+	id, err := uuid.NewV4()
+	if err != nil {
+		return "unknown"
+	}
+	return id.String()
+}
+
+func (s *CerealService) EnqueueWorkflow(ctx context.Context, req *cereal.EnqueueWorkflowRequest) (*cereal.EnqueueWorkflowResponse, error) {
+	logctx := logrus.WithFields(logrus.Fields{
+		"id":            generateRequestID(),
+		"method":        "EnqueueWorkflow",
+		"workflow_name": req.WorkflowName,
+		"instance_name": req.InstanceName,
+	})
+	logctx.Info("enqueuing workflow")
+	if err := s.backend.EnqueueWorkflow(ctx, &backend.WorkflowInstance{
+		InstanceName: req.InstanceName,
+		WorkflowName: req.WorkflowName,
+		Parameters:   req.Parameters,
+	}); err != nil {
+		logctx.WithError(err).Error("failed to enqueue workflow")
+		if err == libcereal.ErrWorkflowInstanceExists {
```

Comment:
they don't have a meaning in this context

Reply:
Why? I'm not sure I understand why we wouldn't return any other error encountered.

Reply:
Here's a list of the errors:
```
	ErrNoTasks                  = errors.New("no tasks in queue")
	ErrNoWorkflowInstances      = errors.New("no workflow instances in queue")
	ErrWorkflowScheduleExists   = errors.New("workflow schedule already exists")
	ErrWorkflowInstanceExists   = errors.New("workflow instance already exists")
	ErrNoDueWorkflows           = errors.New("no due workflows")
	ErrNoScheduledWorkflows     = errors.New("no workflows are scheduled")
	ErrInvalidSchedule          = errors.New("workflow schedule is not valid")
	ErrWorkflowInstanceNotFound = errors.New("workflow instance not found")
	ErrWorkflowScheduleNotFound = errors.New("workflow schedule not found")
	ErrWorkflowNotComplete      = errors.New("workflow instance is still running")
	ErrTaskLost                 = errors.New("task lost before reporting its status")
```
ErrNoTasks is only returned when polling for tasks
ErrNoWOrkflowInstances is returned when polling for workflow instances
ErrWorkflowScheduleExists is returned when trying to put in a schedule
ErrNoDueWorkflows is used in the context of scheduling workflow instances
ErrNoScheduledWorkflows is used in the context of scheduling workflow instances (this can probably be folded up into another error)
ErrInvalidSchedule is used in the context of creating schedules (i guess scheduling too if you messed with the database)
ErrWorkflowInstanceNotFound is only returned when you're looking for a workflow instance
ErrWorkflowScheduleNotFound is returned when you're looking for a schedule
ErrWorkflowNotComplete is used when you try to get the result of an incomplete workflow instance
ErrTaskLost is only used in the context of processing tasks

Reply:
oh wait, i think i see what you're saying